hazelcast/src/main/java/com/hazelcast/map/impl/mapstore/writebehind/ReservedCapacityCounter.java

```diff
@@ -0,0 +1,102 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.map.impl.mapstore.writebehind;
+
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+import static com.hazelcast.internal.util.MapUtil.isNullOrEmpty;
+
+/**
+ * @see WbqCapacityCounter
+ */
+public class ReservedCapacityCounter implements WbqCapacityCounter {
+
+    final ConcurrentMap<UUID, Long> reservedCapacityCountByTxId;
+    final NodeWideUsedCapacityCounter nodeWideUsedCapacityCounter;
+
+    public ReservedCapacityCounter(NodeWideUsedCapacityCounter nodeWideUsedCapacityCounter) {
+        this.nodeWideUsedCapacityCounter = nodeWideUsedCapacityCounter;
+        this.reservedCapacityCountByTxId = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void increment(UUID txnId, boolean backup) {
+        reservedCapacityCountByTxId.compute(txnId, (ignored, currentCapacityCount) -> {
+            if (backup) {
+                nodeWideUsedCapacityCounter.add(1L);
+            } else {
+                nodeWideUsedCapacityCounter.addCapacityOrThrowException(1);
+            }
+            return currentCapacityCount == null ? 1L : (currentCapacityCount + 1L);
+        });
+    }
+
+    @Override
+    public void decrement(UUID txnId) {
+        decrement0(txnId, true);
+    }
+
+    private void decrement0(UUID txnId, boolean decrementNodeWideCounter) {
+        reservedCapacityCountByTxId.computeIfPresent(txnId, (ignored, currentCapacityCount) -> {
+            if (decrementNodeWideCounter) {
+                nodeWideUsedCapacityCounter.add(-1L);
+            }
+            return currentCapacityCount == 1L ? null : currentCapacityCount - 1L;
+        });
+    }
+
+    @Override
+    public void decrementOnlyReserved(UUID txnId) {
+        decrement0(txnId, false);
+    }
+
+    @Override
+    public boolean hasReservedCapacity(UUID txnId) {
+        if (txnId == null) {
+            return false;
+        }
+
+        return reservedCapacityCountByTxId.containsKey(txnId);
+    }
+
+    @Override
+    public Map<UUID, Long> getReservedCapacityCountPerTxnId() {
+        return reservedCapacityCountByTxId;
+    }
+
+    @Override
+    public void increment(Map<UUID, Long> migratedReservedCapacityCountByTxId) {
+        if (isNullOrEmpty(migratedReservedCapacityCountByTxId)) {
+            return;
+        }
+
+        for (Long count : migratedReservedCapacityCountByTxId.values()) {
+            nodeWideUsedCapacityCounter.add(count);
+        }
+        reservedCapacityCountByTxId.putAll(migratedReservedCapacityCountByTxId);
+    }
+
+    @Override
+    public void release() {
+        for (Long counter : reservedCapacityCountByTxId.values()) {
+            nodeWideUsedCapacityCounter.add(-counter);
+        }
```

Comment:
Do we need to call `reservedCapacityCountByTxId.clear()` here?

Reply:
good catch, fixed.